server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt

```diff
@@ -0,0 +1,282 @@
+                            else -> null
+                        }
+                    }
+            } else {
+                null
+            }
+        }
+        .flatten()
+
+private fun ClassDescriptor.canBeExtended() = this.kind.isInterface ||
+    this.modality == Modality.ABSTRACT ||
+    this.modality == Modality.OPEN
+
+private fun FunctionDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
+
+private fun PropertyDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
+            
+// interfaces are ClassDescriptors by default. When calling AbstractClass super methods, we get a ClassConstructorDescriptor
+fun getClassDescriptor(descriptor: DeclarationDescriptor?): ClassDescriptor? =
+        if (descriptor is ClassDescriptor) {
+            descriptor
+        } else if (descriptor is ClassConstructorDescriptor) {
+            descriptor.containingDeclaration
+        } else {
+            null
+        }
+
+fun getSuperClassTypeProjections(
+        file: CompiledFile,
+        superType: KtSuperTypeListEntry
+): List<TypeProjection> =
+        superType
+                .typeReference
+                ?.typeElement
+                ?.children
+                ?.filter { it is KtTypeArgumentList }
+                ?.flatMap { (it as KtTypeArgumentList).arguments }
+                ?.mapNotNull {
+                    (file.referenceExpressionAtPoint(it?.startOffset ?: 0)?.second as?
+                                    ClassDescriptor)
+                            ?.defaultType?.asTypeProjection()
+                }
+                ?: emptyList()
+
+// Checks if the class overrides the given declaration
+fun overridesDeclaration(kotlinClass: KtClass, descriptor: FunctionDescriptor): Boolean =
+        kotlinClass.declarations.any {
+            if (it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
+            ) {
+                if (it is KtNamedFunction) {
+                    parametersMatch(it, descriptor)
+                } else {
+                    true
+                }
+            } else {
+                false
+            }
+        }
+
+fun overridesDeclaration(kotlinClass: KtClass, descriptor: PropertyDescriptor): Boolean =
+        kotlinClass.declarations.any {
+            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
+        }
+
+// Checks if two functions have matching parameters
+private fun parametersMatch(
+        function: KtNamedFunction,
+        functionDescriptor: FunctionDescriptor
+): Boolean {
```

Comment:
Perhaps we could rewrite this into a big functional `zip`? I haven't checked on the specifics though, maybe there's too much specialized logic to do that.

Reply:
A bit unsure about this. We loop through two different lists; value parameters and type parameters. Unsure on how much it would improve in this case, but I may miss something...